### ISSUE TITLE
Move the cranelift-preopt crate content into codegen?

### ISSUE THREAD

**Issue by bnjbvr** [CraneStation/cranelift#767]
I think cranelift-preopt was separated from cranelift-codegen because we assume that, in general, the input IR has been generated by an optimizing compiler (namely, one that ran constant folding). This being said, in our use case, when translating wasm to clif, we're likely to introduce redundancies and constant-foldable instructions that aren't obvious to catch (because we're operating on one wasm binary opcode at a time). There was also another instance of constant-foldable operations being emitted by legalization recently.

In principle, every pass that takes IR and augments it (by adding more IR nodes) is likely to introduce redundancies / constant-foldable instructions. I'd like someone to get more data on this, but how would people feel about putting the cranelift-preopt constant folding facilities into codegen instead?

**Comment by bjorn3** [comment]
+1 Maybe add a setting for enabling and disabling it.

**Comment by sunfishcode** [comment]
The vision here was that we'd add the things that come up when translating optimized wasm in cranelift-codegen/src/simple_preopt.rs, to preserve a simple and relatively fast path for optimized wasm (perhaps eventually paired with a fast register allocator), and then cranelift-preopt is the whole watermelon for non-optimized wasm or other inputs. It should be easy to find things that simple_preopt is missing; just run cranelift-preopt after simple_preopt on some optimized wasm and see what it finds.

And for things like eg. `iadd_imm` instructions, code which produces IR should check for a 0 immediate and avoid emitting instructions that need to be optimized away. Eventually this should be handled automatically by utility functions.

That's the theory anyway. I'd be interested what people think of this design.

**Comment by sunfishcode** [comment]
In any case, +1 to making it easier to run cranelift-preopt's optimizations, wherever it lives.

**Comment by bnjbvr** [comment]
> And for things like eg. iadd_imm instructions, code which produces IR should check for a 0 immediate and avoid emitting instructions that need to be optimized away. Eventually this should be handled automatically by utility functions.

I agree that the best way to reduce the number of IR nodes is not emitting them at all whenever we can :). We don't have the utility functions yet, so in the meanwhile we could make use of the preopt crate.

It seems reasonable to ask users to use the preopt crate in the meanwhile; I just wonder if it is easy to discover, as it is external to the compiler. There are two independent problems here:

- make the code available at compile time:
  - currently it's via an external crate, because of the potential code size hit this could provoke in a build. There's only one optimization (constant folding) at this time, but there could be more (with algebraic identities, etc.).
  - it could be a Cargo feature in codegen, making it easier to discover.
  - we could take the code size hit; arguably, there are passes in Cranelift that could be also externalized (and deferred to the user).
- make the code available at run time:
  - could always be enabled
  - could be enabled thanks to the `OptLevel` setting
  - could get its own setting.

Right now, I'm leaning towards adding it as a Cargo feature in codegen (for discoverability), and then making it conditional on the `OptLevel::Best` optimization level. It could be run as part of the codegen's preopt pass. That could be the first step.

There's another completely different path where we could run constant folding as part of GVN, reducing the number of graph iterations we have to do. (Maybe it could also get more folding opportunities, as it's a global optimization? I don't know!)

Having the InstBuilder default implementation remove redundant operations would be a nice addition and it can be implemented independently.
